tools/tcg_rim_tool/src/main/java/hirs/swid/CredentialParser.java

```diff
@@ -0,0 +1,190 @@
+    private static final String PKCS1_FOOTER = "-----END RSA PRIVATE KEY-----";
+    private static final String PKCS8_HEADER = "-----BEGIN PRIVATE KEY-----";
+    private static final String PKCS8_FOOTER = "-----END PRIVATE KEY-----";
+    private X509Certificate certificate;
+    private PrivateKey privateKey;
+    private PublicKey publicKey;
+
+    public X509Certificate getCertificate() {
+        return certificate;
+    }
+
+    public PrivateKey getPrivateKey() {
+        return privateKey;
+    }
+
+    public PublicKey getPublicKey() {
+        return publicKey;
+    }
+
+    public void parseJKSCredentials() {
+        KeyStore.PrivateKeyEntry privateKeyEntry =
+                parseKeystorePrivateKey(SwidTagConstants.DEFAULT_KEYSTORE_PATH,
+                        SwidTagConstants.DEFAULT_PRIVATE_KEY_ALIAS,
+                        SwidTagConstants.DEFAULT_KEYSTORE_PASSWORD);
+        certificate = (X509Certificate) privateKeyEntry.getCertificate();
+        privateKey = privateKeyEntry.getPrivateKey();
+        publicKey = certificate.getPublicKey();
+    }
+
+    public void parsePEMCredentials(String certificateFile, String privateKeyFile) throws FileNotFoundException {
+        certificate = parsePEMCertificate(certificateFile);
+
+        /*User input on algorithm???*/
+        privateKey = parsePEMPrivateKey(privateKeyFile, "RSA");
+
+        publicKey = certificate.getPublicKey();
+    }
+
+    /**
+     * This method returns the X509Certificate found in a PEM file.
+     * @param filename
+     * @return
+     * @throws FileNotFoundException
+     */
+    private X509Certificate parsePEMCertificate(String filename) throws FileNotFoundException {
+        X509Certificate certificate = null;
+        try {
+            FileInputStream fis = new FileInputStream(filename);
+            BufferedInputStream bis = new BufferedInputStream(fis);
+            CertificateFactory certificateFactory = CertificateFactory.getInstance(X509);
+
+            while (bis.available() > 0) {
+                certificate = (X509Certificate) certificateFactory.generateCertificate(bis);
+            }
+
+
+        } catch (CertificateException e) {
+            System.out.println("Error in certificate factory: " + e.getMessage());
+        } catch (IOException e) {
+            System.out.println("Error reading from input stream: " + e.getMessage());
+        }
+
+        return certificate;
+    }
+
+    /**
+     * This method extracts the private key from a PEM file.
+     * Both PKCS1 and PKCS8 formats are handled.
+     * Algorithm argument is present to allow handling of multiple encryption algorithms,
+     * but for now it is always RSA.
+     * @param filename
+     * @return
+     */
+    private PrivateKey parsePEMPrivateKey(String filename, String algorithm) {
+        PrivateKey privateKey = null;
+        try {
+            File file = new File(filename);
+            FileInputStream fis = new FileInputStream(file);
```

Comment:
It doesn't look like this stream is closed.

Reply:
Added a finally {...} block to close streams.